src/filter.rs

```diff
@@ -0,0 +1,181 @@
+use crate::Node;
+use accesskit::Role;
+use std::fmt::{Debug, Formatter};
+
+/// Create an empty filter.
+/// Convenience function for [`By::new`].
+pub fn by<'a>() -> By<'a> {
+    By::new()
+}
+
+/// A filter for nodes.
+/// The filters are combined with a logical AND.
+pub struct By<'a> {
+    name: Option<&'a str>,
+    name_contains: bool,
+    include_labels: bool,
+    #[allow(clippy::type_complexity)]
+    predicate: Option<Box<dyn Fn(&Node<'_>) -> bool + 'a>>,
+    had_predicate: bool,
+    role: Option<Role>,
+    value: Option<&'a str>,
+}
+
+impl<'a> Debug for By<'a> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let By {
+            name,
+            name_contains,
+            include_labels,
+            predicate,
+            had_predicate,
+            role,
+            value,
+        } = self;
+        let mut s = f.debug_struct("By");
+        if let Some(name) = name {
+            if *name_contains {
+                s.field("name_contains", &name);
+            } else {
+                s.field("name", &name);
+            }
+        }
+        if *include_labels {
+            s.field("include_labels", &true);
+        }
+        if predicate.is_some() || *had_predicate {
+            s.field("predicate", &"<function>");
+        }
+        if let Some(role) = role {
+            s.field("role", &role);
+        }
+        if let Some(value) = value {
+            s.field("value", &value);
+        }
+        s.finish()
+    }
+}
+
+impl<'a> Default for By<'a> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<'a> By<'a> {
+    /// Create an empty filter.
+    pub fn new() -> Self {
+        Self {
+            name: None,
+            name_contains: false,
+            include_labels: false,
+            predicate: None,
+            had_predicate: false,
+            role: None,
+            value: None,
+        }
+    }
+
+    /// Filter by the name of the node with an exact match.
+    pub fn name(mut self, name: &'a str) -> Self {
+        self.name = Some(name);
+        self
+    }
+
+    /// Filter by the name of the node with a substring match.
+    pub fn name_contains(mut self, name: &'a str) -> Self {
+        self.name = Some(name);
+        self.name_contains = true;
+        self
+    }
+
+    /// If a node is labelled by another node, should the label node be included in the results?
+    /// Default is false.
+    pub fn include_labels(mut self) -> Self {
+        self.include_labels = true;
+        self
+    }
+
+    /// Filter by a custom predicate.
+    pub fn predicate(mut self, predicate: impl Fn(&Node<'_>) -> bool + 'a) -> Self {
+        self.predicate = Some(Box::new(predicate));
+        self.had_predicate = true;
```

Comment:
Why is `had_predicate` needed? Isn't it the same as `self.predicate.is_some()`?

Reply:
It's used in By::debug_clone. This way I don't have to clone the predicate function (which wouldn't be useful in debug prints anyways) but can tell the user that the filter was filtered by a custom predicate.